### ISSUE TITLE
Visualization of scene flow and speed estimation

### ISSUE THREAD

**Issue by poornimajd** [visinf/self-mono-sf#2]
Hello great work @hurjunhwa and team !
I was able to run the test code and got the disparity and optical flow images for kitti dataset.I wanted to visualize the scene flow.How do I do that using the disparity and flow? 
Also how to save the 3d motion vector for every test image?
I want to predict the speed of the objects using the 3d motion vector given by the scene flow.To do so,
I just need to take the difference of the two 3d positions in two consecutive frames and divide it by the time interval right?
Does the predicted motion vector require any further scaling?
Any suggestion is greatly appreciated!
Thank you.

**Comment by rshivansh** [comment]
@hurjunhwa yes could you please help with the visualization of scene flow. exact code would be really helpful.

**Comment by hurjunhwa** [comment]
@rshivansh 
Hi, I just pushed the visualization code of scene flow:
https://github.com/visinf/self-mono-sf/tree/master/demo/demo_generator
In the source code, you can configure the directories for source images and output images.

**Comment by poornimajd** [comment]
@hurjunhwa Thanks for the update.Is there a code which saves the 3d motion vector for every pixel in the image?

**Comment by hurjunhwa** [comment]
@poornimajd 
Hi, the visualization code is updated in the following directory. 
https://github.com/visinf/self-mono-sf/tree/master/demo/demo_generator

Yes, exactly. You can divide the 3D motion vector by the time interval. 
However, please note that the 3D motion vector is not defined in the world coordinate, but is a relative 3D motion with respect to the camera. 
Also, the output scale follows the KITTI intrinsics. It needs to be rescaled if you use other datasets with different intrinsics.

**Comment by hurjunhwa** [comment]
The output of the network is 3D motion vector, so you can just use that..
Or, please refer this line in the ```losses.py```, which is what you are looking for. 
https://github.com/visinf/self-mono-sf/blob/5f4e07955351658fa0060e6ecadca6167693b09d/losses.py#L481

We didn't explicitly save the 3D vector but instead convert into KITTI format (2 disparities + flow).
However, it seems also good to have the code for saving the 3D output as well :) 
Thanks for the comment!

**Comment by poornimajd** [comment]
Thanks for the quick reply!

**Comment by rshivansh** [comment]
> @rshivansh
> Hi, I just pushed the visualization code of scene flow:
> https://github.com/visinf/self-mono-sf/tree/master/demo/demo_generator
> In the source code, you can configure the directories for source images and output images.

Hi thanks for the response, visualization looks good, how do you show the visualization for 't+1' frame? I see you have included that in the gif too. I mean I got output for 't -> t+1' but how do you add the second output to this which you show in the demo gif.

**Comment by hurjunhwa** [comment]
Hi,
I am sorry but I didn't fully understand.
The last frame of the visualization is the translated point cloud (at t+1) of points of the source image (at t) using the estimated scene flow. 
Do you mean by visualizing the points at 't+1' using the input frame at 't+1' and the estimated depth at 't+1'?

**Comment by rshivansh** [comment]
Yes, I mean how do you obtain the translated point cloud (at t+1). Also another thing I noticed now is, if I use the ground truth disp and ground truth flow to visualize the outputs look bad. Does the inputs in disp0, disp1, and flow folder have to be in some specific format?

**Comment by hurjunhwa** [comment]
https://github.com/visinf/self-mono-sf/blob/5f4e07955351658fa0060e6ecadca6167693b09d/demo/demo_generator/run.py#L97

This line of the code translates the point cloud at t (``pts1``) using the estimated 3D scene flow (``sf``), here ``tt`` is the discrete time step for visualization between 0 and 1.

The KITTI ground truth flow (``flow``) is the optical flow ground truth, not the scene flow. 
So, that's why it would have looked bad.
These are what they store:
- disp0: disparity map at time t
- disp1: disparity changes to time t+1 (it's not the disparity map at time t+1)
- flow: optical flow from time t to t+1

To obtain ground truth scene flow, you need to manually convert  ``disp0``, ``disp1``, and ``flow`` to scene flow.
You can refer here, which does the inverse calculation (scene flow -> ``disp1`` and ``flow``).
https://github.com/visinf/self-mono-sf/blob/5f4e07955351658fa0060e6ecadca6167693b09d/losses.py#L459

**Comment by rshivansh** [comment]
No I mean in your code I use your image as inputs then I try for another sequence "000012",  for which I obtain the inputs from ground truth disp_0,disp_1, and flow folder. The output does not look like the one I obtianed for "139" from your inputs. Do you know what might be wrong? 
![000012_00](https://user-images.githubusercontent.com/33856562/85307538-06221f80-b47e-11ea-9e38-b2cbe2bcb892.png)
![000139_00](https://user-images.githubusercontent.com/33856562/85307543-07534c80-b47e-11ea-83f8-8101a7352f62.png)

Above one is for 12 and lower for 139

**Comment by hurjunhwa** [comment]
Oh, it seems like the disparity map for 12 is not correctly loaded or differently scaled. 
This is what I got though..

![000012_00](https://user-images.githubusercontent.com/17178388/85308267-4c44a680-b4b1-11ea-8fbf-66a5dbd28381.png)

Did you put the correct disp0? could you maybe show here?

**Comment by rshivansh** [comment]
> Oh, it seems like the disparity map for 12 is not correctly loaded or differently scaled.
> This is what I got though..
> 
> ![000012_00](https://user-images.githubusercontent.com/17178388/85308267-4c44a680-b4b1-11ea-8fbf-66a5dbd28381.png)
> 
> Did you put the correct disp0? could you maybe show here?

Okay this is flow from grounf truth
![000012_10](https://user-images.githubusercontent.com/33856562/85309029-01f70180-b480-11ea-8af0-5b168b898231.png)

This is disp_0 from ground truth
![000012_10](https://user-images.githubusercontent.com/33856562/85309170-336fcd00-b480-11ea-9bae-6ac794e35df8.png)


This is disp_1 from ground truth

![000012_10](https://user-images.githubusercontent.com/33856562/85309205-4387ac80-b480-11ea-9795-d3cb8c1a3a23.png)

and this is vis which is more weird now

![000012_00](https://user-images.githubusercontent.com/33856562/85309245-51d5c880-b480-11ea-9d82-03a255a7d553.png)

Earlier vis was corresponding to my own generated disp and of.

Thanks for quick response.

**Comment by hurjunhwa** [comment]
![000012_00](https://user-images.githubusercontent.com/17178388/85309279-a85bfa80-b4b2-11ea-8dec-f03f9e8b54cd.png)

This is what I got using ground truth.
Maybe there is something wrong when loading the KITTI ground truth images..
which version of pypng do you use? My one is 0.0.20.
Thanks for reporting this malfunctioning :)

**Comment by rshivansh** [comment]
> ![000012_00](https://user-images.githubusercontent.com/17178388/85309279-a85bfa80-b4b2-11ea-8dec-f03f9e8b54cd.png)
> 
> This is what I got using ground truth.
> Maybe there is something wrong when loading the KITTI ground truth images..
> which version of pypng do you use? My one is 0.0.20.
> Thanks for reporting this malfunctioning :)

Hi even I use 0.0.20, also are my inputs correct according to your algo?

**Comment by hurjunhwa** [comment]
Hmm.. that's really odd. I don't have a clear idea for now..
I will think about what can possibly be wrong.

**Comment by rshivansh** [comment]
> Hmm.. that's really odd. I don't have a clear idea for now..
> I will think about what can possibly be wrong.

sure thank you

**Comment by hurjunhwa** [comment]
Yes, your input seems correct. 
As long as the sequence 139 is working in your computer, I guess maybe it's related to data format or data loader..

**Comment by rshivansh** [comment]
> Yes, your input seems correct.
> As long as the sequence 139 is working in your computer, I guess maybe it's related to data format or data loader..

Yes even I think so, however i am unsure of what specific data format your inputs were corresponding to.

**Comment by rshivansh** [comment]
> ![000012_00](https://user-images.githubusercontent.com/17178388/85309279-a85bfa80-b4b2-11ea-8dec-f03f9e8b54cd.png)
> 
> This is what I got using ground truth.
> Maybe there is something wrong when loading the KITTI ground truth images..
> which version of pypng do you use? My one is 0.0.20.
> Thanks for reporting this malfunctioning :)

It is confusing even when you use same inputs (from kitti ground truth) why do we get different outputs, I tried for both versions of open 3d (0.10 and 0.7). The cam_pose.json remains the same right?

**Comment by rshivansh** [comment]
> > ![000012_00](https://user-images.githubusercontent.com/17178388/85309279-a85bfa80-b4b2-11ea-8dec-f03f9e8b54cd.png)
> > This is what I got using ground truth.
> > Maybe there is something wrong when loading the KITTI ground truth images..
> > which version of pypng do you use? My one is 0.0.20.
> > Thanks for reporting this malfunctioning :)
> 
> It is confusing even when you use same inputs (from kitti ground truth) why do we get different outputs, I tried for both versions of open 3d (0.10 and 0.7)

The cam_pose.json is the same right?

**Comment by hurjunhwa** [comment]
Oops, I used pypng **'0.0.18'** and open3d '0.8.0.0'.
Could you maybe downgrade pypng to 0.0.18?

I just remembered that my code needed to be revised for the version 0.0.20, so I decided to just temporarily go for 0.0.18 but didn't change after that.

If you want to keep the version 0.0.20, ``read_png_flow`` and ``read_png_depth`` should be revised accordingly.. which I didn't look into that.

**Comment by rshivansh** [comment]
I downgraded pypng to 0.0.18 and open3d to 0.8.0.0, and this is what I get
![000012_00](https://user-images.githubusercontent.com/33856562/85314798-fb6c8800-b487-11ea-8fac-53483beec1ad.png)

**Comment by hurjunhwa** [comment]
Oh no :( Then I really have no idea.
Yes, cam_pose.json remains the same.
Could you maybe send your GTs or Output (that you want to test) in a zip file and send it to junhwa.hur@visinf.tu-darmstadt.de ?

**Comment by rshivansh** [comment]
Sent

**Comment by rshivansh** [comment]
Great it works now! Thanks for your help @hurjunhwa.

**Comment by rshivansh** [comment]
Hi @hurjunhwa So I am saving the point cloud that is returned by your demo code and would like to segment the moving object in the point cloud. Since the point cloud has the 3D motion vector, I am trying to do if np.sqrt(mvx^2+mvy^2) > Thresholsd, then retain the point cloud points or else make the motion vector 0 (thus delete the points). However I notice that when I do this I get the point cloud of the background and when I do the opposite that is "less than threshold" condition I successfully get the moving object segmnetation. Same is the case with all sequences I tried. Do you have any suggestions to why am I getting the reverse conditions? I tried the same with your predicted results and the same issue is coming. 
So basically my query is in the point cloud the 3D motion vector magnitude of stationary points > moving points. Could you please let me know why that could be happening?
Thank you

**Comment by doublexushuang** [comment]
@hurjunhwa 
Hi ! Thanks for your great work for the visualization code of scene flow!
But it seems that I've got one problem with it.
![image](https://user-images.githubusercontent.com/81733735/117531714-75611d80-b016-11eb-98ee-3be256c708bc.png)
As you may see in the picture, I didn't change any of the data except for the three directories:kitti_data_dir, result_dir, vis_dir
I want to know where the problem may occur.
Is it the path problem or is the installation version of open3d different?
